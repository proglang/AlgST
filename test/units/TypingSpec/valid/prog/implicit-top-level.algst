calc : Int -> Int

{number} : Int
{number} = 42

check : Int
check = calc (implicit[Int] + 10)

synth : Int
synth =
  let r = calc (implicit[Int] + 10) in 
  r
