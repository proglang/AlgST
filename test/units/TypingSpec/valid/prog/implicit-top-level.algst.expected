calc#M.3
|
`- SignatureDecl
   |
   `- type
      |
      `- Type.Arrow U
         |
         +- Type.Type
         |  |
         |  `- TypeRef
         |     |
         |     `- Int
         |
         `- Type.Type
            |
            `- TypeRef
               |
               `- Int

check#M.0
|
`- ValueDecl
   |
   +- type
   |  |
   |  `- Type.Type
   |     |
   |     `- TypeRef
   |        |
   |        `- Int
   |
   +- params
   |
   `- definition
      |
      `- Exp.App
         |
         +- Exp.Abs L implicit [Int] + 10#M.5
         |  |
         |  +- Type.Type
         |  |  |
         |  |  `- TypeRef
         |  |     |
         |  |     `- Int
         |  |
         |  `- Exp.App
         |     |
         |     +- Exp.Var calc#M.3
         |     |
         |     `- Exp.Var implicit [Int] + 10#M.5
         |
         `- Exp.App
            |
            +- Exp.App
            |  |
            |  +- Exp.Var (+)#Builtin.19
            |  |
            |  `- Exp.App
            |     |
            |     +- Exp.TypeApp
            |     |  |
            |     |  +- Exp.Var implicit#Builtin.32
            |     |  |
            |     |  `- Type.Type
            |     |     |
            |     |     `- TypeRef
            |     |        |
            |     |        `- Int
            |     |
            |     `- Exp.Var number#M.1
            |
            `- Exp.Lit
               |
               `- Lit.Int 10

number#M.1
|
`- ValueDecl
   |
   +- implicit
   |
   +- type
   |  |
   |  `- Type.Type
   |     |
   |     `- TypeRef
   |        |
   |        `- Int
   |
   +- params
   |
   `- definition
      |
      `- Exp.Lit
         |
         `- Lit.Int 42

synth#M.2
|
`- ValueDecl
   |
   +- type
   |  |
   |  `- Type.Type
   |     |
   |     `- TypeRef
   |        |
   |        `- Int
   |
   +- params
   |
   `- definition
      |
      `- Exp.Exp
         |
         `- TcExpX.ValueCase
            |
            +- Exp.App
            |  |
            |  +- Exp.Var calc#M.3
            |  |
            |  `- Exp.App
            |     |
            |     +- Exp.Abs L 10#M.6
            |     |  |
            |     |  +- Type.Type
            |     |  |  |
            |     |  |  `- TypeRef
            |     |  |     |
            |     |  |     `- Int
            |     |  |
            |     |  `- Exp.App
            |     |     |
            |     |     +- Exp.App
            |     |     |  |
            |     |     |  +- Exp.Var (+)#Builtin.19
            |     |     |  |
            |     |     |  `- Exp.App
            |     |     |     |
            |     |     |     +- Exp.TypeApp
            |     |     |     |  |
            |     |     |     |  +- Exp.Var implicit#Builtin.32
            |     |     |     |  |
            |     |     |     |  `- Type.Type
            |     |     |     |     |
            |     |     |     |     `- TypeRef
            |     |     |     |        |
            |     |     |     |        `- Int
            |     |     |     |
            |     |     |     `- Exp.Var number#M.1
            |     |     |
            |     |     `- Exp.Var 10#M.6
            |     |
            |     `- Exp.Lit
            |        |
            |        `- Lit.Int 10
            |
            `- r#M.4
               |
               `- Exp.Var r#M.4

